netlogo-gui/src/main/app/AbstractTabsPanel.scala

```diff
@@ -0,0 +1,175 @@
+  def getAppFrame() = { workspace.getFrame.asInstanceOf[AppFrame] }
+  def getAppJFrame() = { jframe }
+  var fileManager: FileManager = null
+  var dirtyMonitor: DirtyMonitor = null
+  var currentTab: Component = interfaceTab
+
+  def initManagerMonitor(manager: FileManager, monitor: DirtyMonitor): Unit =  {
+    fileManager = manager
+    dirtyMonitor = monitor
+    assert(fileManager != null && dirtyMonitor != null)
+  }
+
+  override def processMouseMotionEvent(e: MouseEvent) {
+    // do nothing.  mouse moves are for some reason causing doLayout to be called in the tabbed
+    // components on windows and linux (but not Mac) in java 6 it never did this before and I don't
+    // see any reason why it needs to. It's causing flickering in the info tabs on the affected
+    // platforms ev 2/2/09
+  }
+
+  override def requestFocus() = { currentTab.requestFocus() }
+
+  def getCodeTabsOwner(): JTabbedPane = {
+    tabManager.getCodeTabsOwner.asInstanceOf[JTabbedPane]
+  }
+
+  def getTitleAtCombinedIndex(index: Int): String =  {
+    val (tabOwner, tabIndex) = tabManager.ownerAndIndexFromCombinedIndex(index)
+    tabOwner.getTitleAt(tabIndex)
+  }
+
+  def getComponentAtCombinedIndex(index: Int): Component =  {
+    val (tabOwner, tabIndex) = tabManager.ownerAndIndexFromCombinedIndex(index)
+    tabOwner.getComponentAt(tabIndex)
+  }
+
+  def setComponentAtCombinedIndex(index: Int, tab: Component): Unit = {
+    val (tabOwner, tabIndex) = tabManager.ownerAndIndexFromCombinedIndex(index)
+    tabOwner.setComponentAt(tabIndex, tab)
+  }
+
+  def setSelectedIndexPanels(index: Int): Unit =  {
+    val (tabOwner, tabIndex) = tabManager.ownerAndIndexFromCombinedIndex(index)
+    if (tabOwner.isInstanceOf[CodeTabsPanel]) {
+      tabOwner.requestFocus
+      tabOwner.setSelectedIndex(tabIndex)
+    } else {
+      val selectedIndex = tabManager.getSelectedAppTabIndex
+      if (selectedIndex == tabIndex) {
+        tabManager.setSelectedAppTab(-1)
+      }
+        tabManager.setSelectedAppTab(tabIndex)
+      }
+  }
+
+  def setPanelsSelectedComponent(tab: Component): Unit = {
+    val (tabOwner, tabIndex) = tabManager.ownerAndIndexOfTab(tab)
+    if (tabOwner.isInstanceOf[CodeTabsPanel]) {
+      tabOwner.requestFocus
+      tabOwner.setSelectedIndex(tabIndex)
+    } else {
+      val selectedIndex = tabManager.getSelectedAppTabIndex
+      if (selectedIndex == tabIndex) {
+        tabManager.setSelectedAppTab(-1)
+      }
+        tabManager.setSelectedAppTab(tabIndex)
+      }
+  }
+
+  def getIndexOfCodeTab(tab: CodeTab): Int = {
+    val index = getCodeTabsOwner.indexOfComponent(tab)
+    index + tabManager.getAppTabsOwner.getTabCount
+  }
+
+  // Begin methods intended for use outside of NetLogo code, e.g
+  // for use in extensions
+
+  /**
+    * Removes the specified Component from the appropriate JTabbedPane
+    *
+    * @param tab The Component to remove.
+    */
+   def removeTab(tab: Component): Unit = {
```

Comment:
One thing I forget to add - should this and `addNewTab()` both use `ExtendedCodeTab` as well, if they are meant to be the public interface?  Or am I missing something?

Reply:
The issue with replaceTab is that only a tab  of a similar type can be replaced.

You are right that to make everything more bullet-proof they  should work on an ExtendedTab class, and most methods could be made private[app]. 

I think introducing ExtendedTab is a good idea. 
I think it might be better to hold off on making things private.